### ISSUE TITLE
rateLimitExceeded

### ISSUE THREAD

**Issue by walking** [potmat/homebridge-google-nest-sdm#99]
Hi! I get this error every time I try to change the temperature in the Nest Thermostat

Could not execute device command:  {"response":{"config":{"url":"https://smartdevicemanagement.googleapis.com/v1/enterprises/[myProjectId]/devices/[deviceId]:executeCommand","method":"POST","userAgentDirectives":[{"product":"google-api-nodejs-client","version":"5.1.0","comment":"gzip"}],"data":{"command":"sdm.devices.commands.ThermostatTemperatureSetpoint.SetRange","params":{"heatCelsius":22.5,"coolCelsius":25}},"headers":{"x-goog-api-client":"gdcl/5.1.0 gl-node/16.14.2 auth/7.14.1","Accept-Encoding":"gzip","User-Agent":"google-api-nodejs-client/5.1.0 (gzip)","Authorization":"Bearer [Bearer value]","Content-Type":"application/json","Accept":"application/json"},"params":{},"retry":true,"body":"{\"command\":\"sdm.devices.commands.ThermostatTemperatureSetpoint.SetRange\",\"params\":{\"heatCelsius\":22.5,\"coolCelsius\":25}}","responseType":"json","retryConfig":{"currentRetryAttempt":0,"retry":3,"httpMethodsToRetry":["GET","HEAD","PUT","OPTIONS","DELETE"],"noResponseRetries":2,"statusCodesToRetry":[[100,199],[429,429],[500,599]]}},"data":{"error":{"code":429,"message":"Rate limited for the SetThermostatTemperatureRange command for the user.","errors":[{"message":"Rate limited for the SetThermostatTemperatureRange command for the user.","domain":"global","reason":"rateLimitExceeded"}],"status":"RESOURCE_EXHAUSTED"}},"headers":{"alt-svc":"h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000","cache-control":"private","connection":"close","content-encoding":"gzip","content-type":"application/json; charset=UTF-8","date":"Wed, 06 Sep 2023 05:25:25 GMT","server":"ESF","transfer-encoding":"chunked","user-id":"[UserId]","vary":"Origin, X-Origin, Referer","x-content-type-options":"nosniff","x-frame-options":"SAMEORIGIN","x-xss-protection":"0"},"status":429,"statusText":"Too Many Requests","request":{"responseURL":"https://smartdevicemanagement.googleapis.com/v1/enterprises/[ProjectId]/devices/[deviceId]:executeCommand"}},"config":{"url":"https://smartdevicemanagement.googleapis.com/v1/enterprises/[ProjectId]/devices/[deviceId]:executeCommand","method":"POST","userAgentDirectives":[{"product":"google-api-nodejs-client","version":"5.1.0","comment":"gzip"}],"data":{"command":"sdm.devices.commands.ThermostatTemperatureSetpoint.SetRange","params":{"heatCelsius":22.5,"coolCelsius":25}},"headers":{"x-goog-api-client":"gdcl/5.1.0 gl-node/16.14.2 auth/7.14.1","Accept-Encoding":"gzip","User-Agent":"google-api-nodejs-client/5.1.0 (gzip)","Authorization":"Bearer [Bearer]","Content-Type":"application/json","Accept":"application/json"},"params":{},"retry":true,"body":"{\"command\":\"sdm.devices.commands.ThermostatTemperatureSetpoint.SetRange\",\"params\":{\"heatCelsius\":22.5,\"coolCelsius\":25}}","responseType":"json","retryConfig":{"currentRetryAttempt":0,"retry":3,"httpMethodsToRetry":["GET","HEAD","PUT","OPTIONS","DELETE"],"noResponseRetries":2,"statusCodesToRetry":[[100,199],[429,429],[500,599]]}},"code":429,"errors":[{"message":"Rate limited for the SetThermostatTemperatureRange command for the user.","domain":"global","reason":"rateLimitExceeded"}]} Dining Room Thermostat

The values for projectId, deviceId, Bearer or userId have been replaced, the originals look fine.
The account is new and there is no limit reached in the panel.

Thanks in advance!

**Comment by potmat** [comment]
Hi @walking, does this happen with any other action? The request is definitely blocked by Google due to API rate limiting.  Are you doing lots of other stuff? HKSV perhaps?

**Comment by walking** [comment]
Hi! Thanks for your reply. Nope. Just for the thermostat and every time I try to change the temperature. I'm not using any google device service or any other api or functionality from google.
I've tried to setup it several times and always the same...Usually I can only change the temperature once a day. If I move from 70 to 72, it tries to call it for 70.5,71,71.5 and 72. So every time that I try to change the temperature it can be more than 2 or 3 calls. And that's the limit for the range...
Then I check the dashboard and it says that I only made a couple of calls and I didn't reach any limit...

**Comment by potmat** [comment]
Hi @walking, I've been keeping an eye on this at my home and nothing I do seems to cause this behaviour.  Temperature is always set to exactly what I desire, never in steps.  I even upgraded to the "new Home architecture" thinking it might be that, but nope.

Maybe if anyone else has this issue we can get to the bottom of it.

**Comment by epietrowicz** [comment]
Hey @walking @potmat I just configured this plugin and ran into this issue for my Nest thermostat. What info can I provide to help debug this?

**Comment by potmat** [comment]
Hi @epietrowicz, try enabling debug mode and provide some logs when the system is making too many requests.  Include all the requests, not just the error at the end.  Maybe that will shed some light.

**Comment by walking** [comment]
Hi @dadsalleb can you add more information about how you solved it? Thanks!

**Comment by walking** [comment]
How this is completed? We are still having this issue... there is no change in the code to fix it. Could you please leave it open? Thanks!

**Comment by justin-robinson** [comment]
I'm seeing the same thing. I have 2 thermostats and 4 cameras. This plugin is constantly making my NAS run out of memory now. Applying for a google commercial developer account to see if higher rate limits help.

**Comment by potmat** [comment]
Hi guys, despite everything I've tried I've never been able to reproduce this.  If anyone can, and can submit a PR it would be appreciated.

**Comment by walking** [comment]
I still don't know how I can fix it on my own. The error is still present and there is no way to use it with homebridge

**Comment by polymorpher** [comment]
I had the same rate limit error. I have 3 cameras (Google Home indoor wired, indoor/outdoor wireless, and an old wired Nest camera). Most of the time I don't see any snapshot, get no event / motion detection, and streaming has been really slow. I ended up just deleting the plugin and am buying a bunch of new cameras with good HomeKit / Homebridge support. Would love to get the three Google Home / Nest camera working again in HomeKit some day.

**Comment by tylerlittlefield** [comment]
Fresh install following this guide, hit with rate limit. Should we pin this and maybe warn users in the README about it?

Also, there is a PR for it. Can we merge it? Or is there a way for homebridge to use that branch to test? https://github.com/potmat/homebridge-google-nest-sdm/pull/180

**Comment by potmat** [comment]
Hi @tylerlittlefield, the PR you referenced is about limiting HomeKit integration to a single structure, do you think that is why you are getting rate limited (because you have too many structures)?

**Comment by tylerlittlefield** [comment]
That PR also says it adds throttling to the thermostats requests. So I thought it would be related to the rate limit error we are experiencing.

**Comment by potmat** [comment]
@tylerlittlefield: Yes you are right, sorry about that. My concern is that the vast majority of people don't seem to have this issue (myself included) and I would like to know why some people do before merging any changes. How many devices/structures do you have?

**Comment by tylerlittlefield** [comment]
That's fine, was just curious if anyone knew if I can install that branch to my homebridge. Guess I can just npm install it directly instead of using the ui. I'll report back if it fixes it. 

As for devices, like 20 plus lights, Apple TV, thermostat, camera, smart locks. So quite a few.

Update: No luck with that branch, got a different error so that branch needs some additional work it seems. Restored back to main branch, worked for a couple hours, then rate limit. Fortunately, I can still use this plugin to at least monitor temperature, just can't control it.

**Comment by jharris1829** [comment]
If anyone here is still having the issue where the temperature will snap back to what it was previously, I had been seeing that behavior when in HeatCool mode. As well as rate limiting issues. I just submitted a PR with a fix that worked for me.

What I was seeing was that HomeKit makes a request for both the heat and cool numbers individually when the thermostat is changed, even if you only change one of them. So if I changed the heat, it would first send a request for the heat value, and then a second request for the cool value. The way the plugin handled it would fill in the previous value for the heat on the second request. So Nest was actually changing to the new values and then immediately being told to change back to the old values.

Nest only allows 5 device executions per minute so that's why the rate limiting was filling up so fast. Since it was making two calls for every temperature change, if you change both heat and cool right after each other you have already hit the limit.  The PR fixes that by buffering both temperature values and then makes only one api call.

PR is here #207 would love to have other tests to see if it helps. It has been working in my house this week with no issues. Reading through this thread it sounds like there might be multiple issues going on, but in my case this was an improvement and the plugin works great now.

**Comment by ericly8** [comment]
I hope fixes for the rate limiting can be merged into the supported forks. My house has 7 thermostats, so rate limiting kicks in without too much effort.
